CBG-804: Update test OIDC provider for compatibility with updated library

sarathkumarsivan opened the pull request:
auth/auth.go
- Removed unused imports for jose, oidc which are no longer required since the library update.
- Modified the AuthenticateUntrustedJWT signature and return values to return *oidc.IDToken instead of jose.JWT.
- Revised the changes to parse JWT to determine the issuer and provider in AuthenticateUntrustedJWT method.
- Added changes to verify the JWT in AuthenticateUntrustedJWT by using the new library.
- Modified logging statements in AuthenticateUntrustedJWT to accommodate the library upgrade.
- Modified the AuthenticateTrustedJWT signature and return values to return *oidc.IDToken instead of jose.JWT.
- Revised the code in AuthenticateTrustedJWT to verify claims to ensures that the token we received from the provider is valid for SG.
- Changed the token argument while calling authenticateJWT function from AuthenticateTrustedJWT.
- Modified the authenticateJWT signature and return values to return *oidc.IDToken instead of jose.JWT.
-  Removed code the code changes in authenticateJWT to extract identity from JWT; identity claims such as ID, Name, Email, ExpireAt can be directly accessed from *oidc.IDToken passed as input parameter to this method. The claims are already extracted in the caller.
- Argument change in GetOIDCUsername functional for fetching the username with subject claim extracted from JWT.
- Modified logging statements in authenticateJWT to accommodate the library upgrade.
- Modified the code to check whether the email needs to be updated (e.g. user has changed email in external auth system).
- Added changes to extract the Email claim from raw claims available in JWT.
- Auto registration code revision to accommodate the library upgrade.

auth/auth_test.go
- Removed the tests which were obsolete since the library upgrade.
- Revised auth related unit tests to accommodate the new changes.
- Added additional tests to verify the new code changes in auth package.

auth/jwt.go
- Removed GetJWTIdentity function definition to extracts the JWT Identity Claims (includes ID, Email, Expiry) from a JWT; obsolete code.
- Added an unexported fuction to parse JWT and check the claims.
- Removed GetJWTExpiry function definition which returns the "exp" claim (Identity.ExpiresAt) for the JWT, as a time.Time; osbolete code.
- Removed GetJWTIssuer function definition which extracts the issuer and audience from JWT; obsolete code.

auth/jwt_test.go
- Removed the tests which were obsolete since the library upgrade.
- Revised auth relatest unit tests to accommodate the new changes.
- Added additional tests to verify the new code changes in auth package.

auth/oidc.go
- Removed http, oauth2, oidc package imports which are obsolete and added go-oidc, context, oauth2, ad go-jose.v2 package imports which are required.
- Added OIDCClient struct to encapsulates the provider, oauth configuration and context.
- Removed *oidc.Client from OIDCProvider struct and replaced it with OIDCClient.
- The GetClient method signature has been modified to return *OIDCClient instead of *oidc.Client
- Added changes to receive provider config, context, discovery sync flag and error when calling DiscoverConfig to fetch the provider metadata or OpenID Connect discovery configuration.
- Removed changes to call discovery sync from InitOIDCClient; the new library doesn't have the support for this.
- Added changes to set the oauth configuration for oauth requests.
- Modified DiscoverConfig method to support standard issuer based provider config discovery.
- The new library only support provider discovery based on standard issuer based discovery endpoint; so added changes to support discovery from non-standard wellknown endpoint.
- Remoed OIDCToHTTPError function which converts an OpenID Connect / OAuth2 error to an HTTP error; obsolete code.
- Added addition NOT exported methods required for fetching provide configuration.

auth/oidc_provider.
- Removed OidcProviderConfiguration function; obsolete.
- Removed parseURI method; obsolete.
- Removed unused package imports.

auth/oidc_provider_test.go
- Removed the tests which were obsolete since the library upgrade.
- Revised auth related unit tests to accommodate the new changes.
- Added additional tests to verify the new code changes in auth package.

auth/oidc_test.go
- Removed the tests which were obsolete since the library upgrade.
- Revised auth related unit tests to accommodate the new changes.
- Added additional tests to verify the new code changes in auth package.

manifest/default.xml
- Updated the commit hash of latest OIDC library.

rest/handler.go
- Removed the function call OIDCToHTTPError to map OIDC/OAuth2 errors to HTTP form.

rest/oidc_api.go
- Removed obsolete constants, package imports and additional constants required to support the upgrade.
- Changed the data type of Expiry value in OIDCTokenResponse to time.Time from int.
- Formatted code changes in handleOIDC.
- Revised handleOIDCCommon, handleOIDCCallback, handleOIDCRefresh, createSessionForTrustedIdToken, getOIDCCallbackURL functions to support the upgrade.

rest/oidc_test_provider.go
- Added changes to test the OpenID Connect workflow.

rest/oidc_test_provider_test.go
- Added unit tests for provider test.

bbrks commented on manifest/default.xml:
```diff
@@ -80,7 +80,7 @@
   <project name="osext" path="godeps/src/github.com/kardianos/osext" remote="couchbasedeps" revision="29ae4ffbc9a6fe9fb2bc5029050ce6996ea1d3bc"/>
 
-  <project name="go-oidc" path="godeps/src/github.com/coreos/go-oidc" remote="couchbasedeps" revision="5aa9381f6e998aa16cc96b4347d33dcc29792864"/>
+  <project name="go-oidc" path="godeps/src/github.com/coreos/go-oidc" remote="couchbasedeps" revision="8d771559cf6e5111c9b9159810d0e4538e7cdc82"/>
```
All of our third-party dependencies are mirrored here: https://github.com/couchbasedeps/

We'll need to update go-oidc here to include this commit.

Let's run through this in sync-up and put it into the dev knowledge base doc at the same time.

sarathkumarsivan replied:
Thank you Ben for pointing this out. The go-oidc repo has been mirrored and updated to latest revision by running the Jenkins job: http://server.jenkins.couchbase.com/job/update-couchbasedep/85

bbrks commented on rest/oidc_test_provider.go:
```diff
@@ -13,28 +13,37 @@ import (
-	"github.com/coreos/go-oidc/key"
 	"github.com/couchbase/sync_gateway/auth"
 	"github.com/couchbase/sync_gateway/base"
+	"gopkg.in/square/go-jose.v2"
```
We'll need this mirroring into couchbasedeps and including in the manifest, as it's a new dependency.

sarathkumarsivan replied:
Thank you Ben for pointing this out. The go-jose repo has been mirrored and updated to latest revision by running the Jenkins job: http://server.jenkins.couchbase.com/job/update-couchbasedep/84

adamcfraser commented on auth/oidc.go:
```diff
@@ -209,76 +210,168 @@ func (op *OIDCProvider) DiscoverConfig() (config *oidc.ProviderConfig, shouldSyn
+
+type audience []string
+
+func GetIDToken(rawIDToken string) (*oidc.IDToken, error) {
```
Looking further, it seems like this functionality is already implemented in the square/go-jose.v2 library we're importing elsewhere (ParseSigned, UnsafeClaimsWithoutVerification).  Reusing the existing library instead of reimplementing seems like a better approach.

adamcfraser commented on auth/auth.go:
```diff
@@ -412,115 +411,100 @@ func (auth *Authenticator) AuthenticateUser(username string, password string) Us
 
 	base.Debugf(base.KeyAuth, "Call GetProviderForIssuer w/ providers: %+v", base.UD(providers))
-	provider := providers.GetProviderForIssuer(issuer, audiences)
+	provider := providers.GetProviderForIssuer(idToken.Issuer, idToken.Audience)
```
In the case that there's only a single provider defined in providers, the work to extract the issuer and audience from the token manually isn't necessary - we could jump straight to attempted verification using the provider, and let verification fail on the issuer/audience mismatch.  We should refactor to make this the default path (since multiple providers is the uncommon use case), and only do the work to manually extract issuer/audience in the multi-provider case.

adamcfraser commented on auth/auth.go:
```diff
@@ -412,115 +411,100 @@ func (auth *Authenticator) AuthenticateUser(username string, password string) Us
 
-	err = client.VerifyJWT(jwt)
+	config := &oidc.Config{ClientID: *provider.ClientID}
+	idTokenVerifier := provider.OIDCClient.Provider.Verifier(config)
```
As discussed, we'd prefer to create a Verifier once and cache it for use for multiple authentications, as was previously done with the OIDCClient.  Is there a reason that can't be done?

adamcfraser commented on auth/auth.go:
```diff
@@ -412,115 +411,100 @@ func (auth *Authenticator) AuthenticateUser(username string, password string) Us
-		base.Debugf(base.KeyAuth, "Error getting JWT issuer: %v", err)
-		return nil, jose.JWT{}, err
+		base.Debugf(base.KeyAuth, "Error parsing JWT string to IDToken in AuthenticateUntrustedJWT: %v", err)
+		return nil, &oidc.IDToken{}, err
```
In the error cases (here and below), is there a reason to return a pointer to an empty token, instead of just nil?

sarathkumarsivan commented on auth/oidc.go:
```diff
@@ -209,76 +210,168 @@ func (op *OIDCProvider) DiscoverConfig() (config *oidc.ProviderConfig, shouldSyn
+type audience []string
+
+func GetIDToken(rawIDToken string) (*oidc.IDToken, error) {
+	_, err := jose.ParseSigned(rawIDToken)
```
The parseJWT will not validate whether the token is well-formed or not. Instead it just checks for invalid claims. But this can be removed since we have GetJWTIssuer is already in place.

sarathkumarsivan commented on auth/oidc.go:
```diff
@@ -209,76 +210,168 @@ func (op *OIDCProvider) DiscoverConfig() (config *oidc.ProviderConfig, shouldSyn
 
-	base.Debugf(base.KeyAuth, "Fetching custom provider config from %s", base.UD(discoveryURL))
-	req, err := http.NewRequest("GET", discoveryURL, nil)
+	// Throw out tokens with invalid claims before trying to verify the token.
```
This has been removed from latest revision.

sarathkumarsivan commented on auth/oidc.go:
```diff
@@ -209,76 +210,168 @@ func (op *OIDCProvider) DiscoverConfig() (config *oidc.ProviderConfig, shouldSyn
+		return nil, fmt.Errorf("oidc: malformed jwt: %v", err)
+	}
+	var token idToken
+	if err := json.Unmarshal(payload, &token); err != nil {
```
Makes sense, this is fixed in latest revision.

adamcfraser commented on auth/oidc.go:
```diff
@@ -141,65 +174,58 @@ func (op *OIDCProvider) InitOIDCClient() error {
-	config, shouldSyncConfig, err := op.DiscoverConfig()
-	if err != nil || config == nil {
+	ctx := context.Background()
+	provider, _, err := op.DiscoverConfig(ctx)
```
Using the local variable name 'providerConfig' here would be less confusing - currently it sounds like the same thing as OIDCProvider.

adamcfraser commented on auth/oidc.go:
```diff
@@ -141,65 +174,58 @@ func (op *OIDCProvider) InitOIDCClient() error {
-
-		var standardConfig oidc.ProviderConfig
+		wellKnown := strings.TrimSuffix(op.Issuer, "/") + discoveryConfigPath
+		base.Infof(base.KeyAuth, "Fetching provider config from standard issuer-based discovery endpoint: %s", base.UD(wellKnown))
```
It seems unnecessary to build the wellKnown endpoint just for the purposes of logging.  I think it's reasonable to just log the issuer here, if anything.

adamcfraser commented on auth/oidc.go:
```diff
@@ -141,65 +174,58 @@ func (op *OIDCProvider) InitOIDCClient() error {
 
-func (op *OIDCProvider) DiscoverConfig() (config *oidc.ProviderConfig, shouldSync bool, err error) {
-
+func (op *OIDCProvider) DiscoverConfig(ctx context.Context) (config *oidc.Provider, shouldSync bool, err error) {
```
I realize shouldSync isn't currently being used, but when we add that functionality back in, it makes sense to support sync for both standard and explicitly defined discovery endpoints.  So I think it's reasonable to remove from here.

adamcfraser commented on auth/oidc.go:
```diff
@@ -212,73 +238,148 @@ func (op *OIDCProvider) DiscoverConfig() (config *oidc.ProviderConfig, shouldSyn
+// https://github.com/coreos/go-oidc/blob/v2/oidc.go#L113
+func (op *OIDCProvider) FetchCustomProviderConfig(ctx context.Context, discoveryURL string) (*oidc.Provider, error) {
 	// If discovery URL is empty, use the standard discovery URL
 	if discoveryURL == "" {
```
We're already doing this check in DiscoverConfig.  I think we should return error here if an empty string is passed in, instead of calling back to standard.

adamcfraser commented on auth/oidc.go:
```diff
@@ -212,73 +238,148 @@ func (op *OIDCProvider) DiscoverConfig() (config *oidc.ProviderConfig, shouldSyn
+	}
+
+	var p ProviderMetadata
+	err = unmarshalResp(resp, body, &p)
```
Was it necessary to switch from the previous Decode approach?  Adding the status check seems like a reasonable enhancement, but I'm less sure about the other changes in unmarshalResp.  It looks like it's just providing some additional Content-Type error information which doesn't really provide particular value.

adamcfraser commented on auth/oidc.go:
```diff
@@ -212,73 +238,148 @@ func (op *OIDCProvider) DiscoverConfig() (config *oidc.ProviderConfig, shouldSyn
-	base.Debugf(base.KeyAuth, "Returning config: %v", base.UD(oidcConfig))
-	return &oidcConfig, nil
+	provider := &oidc.Provider{}
+	SetUnexportedField(reflect.ValueOf(provider).Elem().FieldByName("issuer"), p.Issuer)
```
Using reflection to set unexported fields is a red flag, as it defeats the purpose of using a library (if we're modifying unexported fields, we can no longer be sure that our usage of the library is supported/valid).

In this case, 'provider' is really just the provider config.  We're not using it outside of the client init, so I think we can just remove as a property on OIDCClient.

We still need the ability to create the idTokenVerifier, but it looks like that could be done using the (public) NewVerifier (cloning some of the functionality in provider.Verifier.

adamcfraser commented on auth/oidc.go:
```diff
@@ -141,65 +174,58 @@ func (op *OIDCProvider) InitOIDCClient() error {
 
-func (op *OIDCProvider) DiscoverConfig() (config *oidc.ProviderConfig, shouldSync bool, err error) {
-
+func (op *OIDCProvider) DiscoverConfig(ctx context.Context) (config *oidc.Provider, shouldSync bool, err error) {
```
It looks like we're only ever passing in the background context here.  If that's the case, what's the value in using it here, and in FetchCustomProviderConfig?

adamcfraser replied:
'the background context per client' doesn't make sense.  There's only a single background context, and it doesn't actually provide any context-type functionality (never cancelled, no deadline, etc). 

In particular, since there's one long-lived OIDCClient per provider that is *not* associated with a given auth request, I don't think it makes sense for OIDCClient to have a Context property.

adamcfraser commented on auth/oidc.go:
```diff
@@ -212,73 +238,148 @@ func (op *OIDCProvider) DiscoverConfig() (config *oidc.ProviderConfig, shouldSyn
 		return nil, err
 	}
-	resp, err := http.DefaultClient.Do(req)
+	resp, err := doRequest(ctx, req)
```
What's the advantage of switching to this, given that this isn't being done within anything other than background context?

adamcfraser commented on auth/oidc.go:
```diff
@@ -212,73 +238,148 @@ func (op *OIDCProvider) DiscoverConfig() (config *oidc.ProviderConfig, shouldSyn
-	return fmt.Sprintf("%s_%s", provider.UserPrefix, url.QueryEscape(subject))
+// Clone of provider discovery request execution from go-oidc library.
+// https://github.com/coreos/go-oidc/blob/8d771559cf6e5111c9b9159810d0e4538e7cdc82/oidc.go#L58
+func doRequest(ctx context.Context, req *http.Request) (*http.Response, error) {
```
As mentioned above, this doesn't seem necessary.

adamcfraser commented on auth/oidc.go:
```diff
@@ -212,73 +238,148 @@ func (op *OIDCProvider) DiscoverConfig() (config *oidc.ProviderConfig, shouldSyn
+	for _, a := range p.IdTokenSigningAlgValuesSupported {
+		if supportedAlgorithms[a] {
+			algs = append(algs, a)
+		}
```
Logging a warning on an unsupported algorithm in the custom provider seems like a good idea.

sarathkumarsivan commented on auth/auth_test.go:
```diff
@@ -647,32 +605,9 @@ func TestAuthenticateTrustedJWT(t *testing.T) {
-	assert.NotNil(t, jws)
-}
-
-func TestAuthenticateJWTWithNoClaim(t *testing.T) {
```
Functions from the old library were used in those removed test cases, and testing logic needs to be corrected as per the new change. Additional unit test can be covered in another PR.

adamcfraser commented on rest/oidc_api.go:
```diff
@@ -103,114 +100,121 @@ func (h *handler) handleOIDCCommon() (redirectURLString string, err error) {
 	}
 
-	tokenResponse, err := oac.RequestToken(oauth2.GrantTypeRefreshToken, refreshToken)
+	token, err := client.Config.TokenSource(client.Context, &oauth2.Token{RefreshToken: refreshToken}).Token()
```
Have you validated the refresh functionality against a live provider?

sarathkumarsivan commented on auth/oidc.go:
```diff
@@ -233,52 +267,94 @@ func (op *OIDCProvider) FetchCustomProviderConfig(discoveryURL string) (*oidc.Pr
-	return fmt.Sprintf("%s_%s", provider.UserPrefix, url.QueryEscape(subject))
+// List of algorithms explicitly supported by go-oidc library.
+// If a provider supports other algorithms, such as HS256 or none,
+// those values won't be passed to the IDTokenVerifier.
```
Comments are added against the cloned code from go-oidc package, also updated the manifest file.

sarathkumarsivan commented on auth/oidc_verify.go:
```diff
@@ -0,0 +1,188 @@
+
+const (
+	issuerGoogleAccounts         = "https://accounts.google.com"
+	issuerGoogleAccountsNoScheme = "accounts.google.com"
```
As per the OpenID Connect specs, providers will always return a signed token.

sarathkumarsivan replied:
[CBG-841](https://issues.couchbase.com/browse/CBG-841) has been filed to track this.

sarathkumarsivan commented on auth/auth_test.go:
```diff
@@ -625,54 +626,17 @@ func TestAuthenticateTrustedJWTWithBadClaim(t *testing.T) {
+	token, err := builder.CompactSerialize()
+	require.NoError(t, err, "Failed to serialize JSON Web Token")
 
-func TestAuthenticateTrustedJWT(t *testing.T) {
```
Added coverage against the new code changes.

adamcfraser commented on auth/auth.go:
```diff
@@ -417,110 +418,116 @@ func (auth *Authenticator) AuthenticateUser(username string, password string) Us
 // If the token is validated but the user for the username defined in the subject claim doesn't exist,
 // creates the user when autoRegister=true.
-func (auth *Authenticator) AuthenticateTrustedJWT(token string, provider *OIDCProvider, callbackURLFunc OIDCCallbackURLFunc) (User, jose.JWT, error) {
+func (auth *Authenticator) AuthenticateTrustedJWT(token string, provider *OIDCProvider, callbackURLFunc OIDCCallbackURLFunc) (User, time.Time, error) {
```
It looks like callbackURLFunc is now unnecessary?

sarathkumarsivan replied:
Yeah, you're right. Method signature of AuthenticateTrustedJWT has been revised by removing the OIDCCallbackURLFunc function argument which is no longer being used.

adamcfraser commented on auth/auth.go:
```diff
@@ -417,110 +418,116 @@ func (auth *Authenticator) AuthenticateUser(username string, password string) Us
 	if userErr != nil {
-		base.Debugf(base.KeyAuth, "Failed to get OIDC User from %v.  Error: %v", base.UD(username), userErr)
-		return nil, jwt, userErr
+		base.Debugf(base.KeyAuth, "User %q doesn't exists, Error: %v", base.UD(username), userErr)
```
```suggestion
		base.Debugf(base.KeyAuth, "Error retrieving user for username %q: %v", base.UD(username), userErr)
```

sarathkumarsivan replied:
Debug level log message has been revised.

adamcfraser commented on auth/oidc.go:
```diff
@@ -141,88 +199,86 @@ func (op *OIDCProvider) InitOIDCClient() error {
 
-func (op *OIDCProvider) DiscoverConfig() (config *oidc.ProviderConfig, shouldSync bool, err error) {
-
+func (op *OIDCProvider) DiscoverConfig(ctx context.Context) (verifier *oidc.IDTokenVerifier, endpoint *oauth2.Endpoint, err error) {
```
It looks like we're still passing the background context down the stack from InitOIDCClient to here.  Since this is just the background context, I think it's sufficient to just pass the background context directly to the context-aware functions being called in DiscoverConfig, and remove the parameter.

sarathkumarsivan replied:
Removed background context parameter from DiscoverConfig and added changes to just pass background context directly to the context-aware function called directly in DiscoverConfig.

adamcfraser commented on rest/oidc_api.go:
```diff
@@ -15,34 +15,38 @@ import (
+	Username     string    `json:"name,omitempty"`          // Sync Gateway user name
+	AccessToken  string    `json:"access_token,omitempty"`  // Access token, from OP
+	TokenType    string    `json:"token_type,omitempty"`    // Access token type, from OP
+	Expires      time.Time `json:"expires_in,omitempty"`    // Access token expiry, from OP
```
This doesn't align with the OIDC spec (and would break backwards compatibility even if it was to spec):
```
expires_in
OPTIONAL. Expiration time of the Access Token in seconds since the response was generated.
```

sarathkumarsivan replied:
Added changes to keep the expiration time of the access token in seconds since the response was generated. Now it is aligned with the OpenID Connect spec and is backward compatible.

sarathkumarsivan commented on auth/auth.go:
```diff
@@ -417,110 +418,116 @@ func (auth *Authenticator) AuthenticateUser(username string, password string) Us
-	if err != nil {
-		base.Debugf(base.KeyAuth, "Error parsing JWT in AuthenticateTrustedJWT: %v", err)
-		return nil, jose.JWT{}, err
+	if provider.ClientID == nil {
```
Makes much sense. Refactored clientID definition from *string to string, added changes to check for empty string instead of nil in NewDatabaseContext, and removed the duplicate check from AuthenticateTrustedJWT.

sarathkumarsivan commented on auth/oidc.go:
```diff
@@ -7,21 +7,39 @@
 //  either express or implied. See the License for the specific language governing permissions
 //  and limitations under the License.
 
+// Since Hydrogen, Sync Gateway started using the latest github.com/coreos/go-oidc package which enables
```
This is change has been included in the latest revision.

sarathkumarsivan commented on auth/oidc.go:
```diff
@@ -7,21 +7,39 @@
+// discovery in Sync Gateway while using the new github. com/coreos/go-oidc package requires cloning certain
+// un-exported functions/methods, variables from  the library with quite a bit of refactoring.
+//
+// parseClaim: https://github.com/coreos/go-oidc/blob/v2/verify.go#L174
```
Comments about the OIDC library has been revised and used Github permalinks.

bbrks commented:
@sarathkumarsivan Would probably be wise to rebase onto master and fix the conflicts introduced by the merged OIDC callback PR #4551 before another review takes place to reduce amount of re-reviewing effort.

sarathkumarsivan commented:
> @sarathkumarsivan Would probably be wise to rebase onto master and fix the conflicts introduced by the merged OIDC callback PR #4551 before another review takes place to reduce amount of re-reviewing effort.

Merge conflict has been resolved and did some cleanup

sarathkumarsivan commented:
CBG-820: Additional coverages with GetRSASigner refactoring

- Refactor TestGetOIDCCallbackURL with end to end coverage.
- Move GetRSASigner to util_testing.go for reusing it across tests in other packages.
- Return error when auto registration is disabled against new user authentication.
- Expose DiscoveryConfigPath to reuse it across rest package
- Additional coverage for the following scenarios:
  - New user auto registration enabled single provider
  - New user auto registration disabled single provider
  - New user auto registration enabled no access token single provider
  - New user auto registration enabled multiple providers
  - New user auto registration disabled multiple providers
  - New user auto registration disabled no access token multiple providers

sarathkumarsivan commented:
- Additional coverage in TestGetOIDCCallbackURL to check whether the provider is getting added to the OIDC callback URL for both oidc and oidc challenge requests. Following scenarios are included:
  - oidc default provider specified but current provider is not default.
  - oidc challenge default provider specified but current provider is not default.
  - oidc default provider specified and current provider is default.
  - oidc default provider specified with no current provider.
  - oidc challenge default provider specified with no current provider.

- Additional coverage in TestOpenIDConnectAuth:
  - Auto registration failure
  - Authorization server sends the callback URL without auth code.
  - OpenID Connect callback received with error code and error description from Auth server.
  - No provider in the callback URL returned from authorization server.

adamcfraser commented on rest/oidc_api_test.go:
```diff
@@ -1,139 +1,612 @@
+var (
+	optionsOnce sync.Once
+	options     mockOptions
+)
```
As was commented earlier, we should avoid global variables unless absolutely needed.  It looks like you've just moved the previous globals (issuerGoogle, issuerFacebook) into this struct.  

Ben already outlined a recommended approach on the other PR (https://github.com/couchbase/sync_gateway/pull/4558/files#r420686763) - we should be doing that here.

sarathkumarsivan replied:
Makes much sense, thank you for clarifying this. I've refactored the mockAuthServer code quite a bit to avoid these global scoped variable.

adamcfraser commented on rest/oidc_api_test.go:
```diff
@@ -1,139 +1,612 @@
+			TokenEndpoint: issuerGoogle + "/token", JwksUri: issuerGoogle + "/oauth2/v3/certs",
+			IdTokenSigningAlgValuesSupported: []string{"RS256"},
+		}
+	} else if provider == "facebook" {
```
If these are just test issuers/providers, is there a good reason to name them 'google' and 'facebook'? Are these testing configurations or token formats specific to those real-world providers? If not, keeping the tests generic would be preferable, IMO.

sarathkumarsivan replied:
Yeah, these are just test issuers and providers and there is no good reason to name them 'google' and 'facebook'; just given this name. These names came to my thought at the time of writing this code and the token formats NOT specific to those real-world providers. I see what you are suggesting, added the changes to make these tests more generic.

adamcfraser commented on rest/oidc_api_test.go:
```diff
@@ -1,139 +1,612 @@
-		require.NoError(t, err, "redirect_uri should be a valid URL")
-		assert.Equal(t, providerSalesforce.Name, redirectURL.Query().Get(authProvider))
-	})
+	type test struct {
```
'wantProviderName' and 'expectedProvider' appear to be redundant. 

inputRequestURL and 'oidcChallenge' are also redundant.  Cleaning these up will make the tests easier to understand.

adamcfraser commented on rest/oidc_api_test.go:
```diff
@@ -1,139 +1,612 @@
+	renderJSON(w, r, http.StatusOK, response)
+}
+
+func TestGetOIDCCallbackURL(t *testing.T) {
```
A comment describing what this test is actually testing would be useful.

adamcfraser commented on rest/oidc_api_test.go:
```diff
@@ -1,139 +1,612 @@
+				resp := rt.SendAdminRequest(http.MethodGet, tc.inputRequestURL, "")
+				require.Equal(t, http.StatusUnauthorized, resp.Code)
+				wwwAuthHeader := resp.Header().Get("Www-Authenticate")
+				assert.Contains(t, wwwAuthHeader, tc.expectedProvider)
```
Shouldn't we be verifying the redirectURI in the challenge case, as we're doing below for the oidc case?

sarathkumarsivan commented on auth/oidc.go:
```diff
@@ -43,7 +43,7 @@ import (
 
 const (
-	discoveryConfigPath = "/.well-known/openid-configuration"
+	DiscoveryConfigPath = "/.well-known/openid-configuration"
```
Makes sense, renamed DiscoveryConfigPath to OIDCDiscoveryConfigPath.

sarathkumarsivan commented:
> Flagged a few things that were already raised on the previous PR. Can you address those and then reassign it back to me for another pass?

Sure Adam, I've made the suggested changes and the latest revision is out the door for another pass.

sarathkumarsivan commented:
- Added coverage for the following scenarios:
  - initiate auth request with not configured provider
  - auth request with discovery failure from OAuth server
  - auth request with access type not offline
  - token exchange error from OAuth server
  - no ID token received from OAuth server
  - token refresh token exchange error from OAuth server
  - token refresh no ID token received from OAuth server
- Add `assertHttpResponse` method to reuse the similar logic which are similar.
- Added `omitempty` property against ID Token field for better coverage.

adamcfraser commented on rest/oidc_api_test.go:
```diff
@@ -1,139 +1,865 @@
+	}
+	tests := []test{
+		{
+			name: "new user auto registration enabled single provider",
```
The test 'name' property in these tests (and others) isn't concise enough for a name, but isn't verbose enough to be a useful description.  For example, on initial scan, it's not obvious whether this is 'auto registration enabled', or 'enabled single provider'.  What do you think about having a more concise name property, and more verbose description property?

adamcfraser commented on rest/oidc_api_test.go:
```diff
@@ -1,139 +1,865 @@
+			wantUsername:       "foo_noah",
+			authURL:            "/db/_oidc?provider=foo&offline=true",
+			register:           true,
+			includeAccessToken: false,
```
Why doesn't this match the value on the provider?  More generally, is this necessary as a parameter, or can it be taken from the provider?

adamcfraser commented on rest/oidc_api.go:
```diff
@@ -103,114 +101,123 @@ func (h *handler) handleOIDCCommon() (redirectURLString string, err error) {
+
+	rawIDToken, ok := token.Extra("id_token").(string)
+	if !ok {
+		return base.HTTPErrorf(http.StatusInternalServerError, "No id_token field in oauth2 token.")
```
Why do you think we should return a StatusInternalServerError here, as opposed to StatusUnauthorized as we're doing above on the failed exchange?

adamcfraser commented on rest/oidc_api_test.go:
```diff
@@ -1,139 +1,865 @@
+			response, err := http.DefaultClient.Do(request)
+			require.NoError(t, err, "Error sending request")
+
+			if !tc.register {
```
As discussed, it's preferable to avoid a long list of conditions like this, when only a single test case is expected to evaluate to true.  In this case, since we're always evaluating the http response for the initial call, a test parameter for the error would be preferable (and would also make it obvious what each test is evaluating from the test case definition).

adamcfraser commented on rest/oidc_api_test.go:
```diff
@@ -1,139 +1,865 @@
+// It is not intended to be used as an actual OAuth 2 server. It lacks many features that would
+// be required in a classic implementation. See https://tools.ietf.org/html/rfc6749 to know more
+// about the OAuth 2.0 Authorization Framework specification.
+type mockAuthServer struct {
```
Can the mockAuthServer tests be run against the previous OIDC implementation?  It would useful to ensure we aren't changing any behaviour that didn't previously have test coverage.
